### DISCUSSION TITLE
Re-using fastapi.Body embedded object with pydantic BaseModel changes param name

### DISCUSSION THREAD

**Discussion by tonyay163** [fastapi/fastapi#11682] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

import fastapi
from pydantic import BaseModel, Field


class Param(BaseModel):
    param: str = Field(default=None)


PARAM = fastapi.Body(
    description="param",
    embed=True,
    default=None,
)


app = fastapi.FastAPI()


@app.post("/one")
async def post_one(one: Optional[Param] = PARAM):
    return one


@app.post("/two")
async def post_two(two: Optional[Param] = PARAM):
    return two
```

### Description

I'm trying to upgrade from 0.99.1 to 0.111.0 which involves the pydantic v2 upgrade. This specific case seems to be breaking in my tests when I try to upgrade, I'm not positive it's related to pydantic, but the problem doesn't seem to be there if I'm not using pydantic request models.

This is correct in Swagger docs (note that if you have pydantic v1 installed, the docs are wrong as well):
![image](https://github.com/tiangolo/fastapi/assets/2197329/a50f52b5-8318-49b2-a570-42a3055d662b)

But a curl changes /two to require "one" instead "two":
```
➜  ~ curl localhost:8000/two -H "content-type: application/json" --data '{"two": {"param": "hello"}}'
null%

➜  ~ curl localhost:8000/two -H "content-type: application/json" --data '{"one": {"param": "hello"}}'
{"param":"hello"}%
```

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.7.3

### Python Version

3.11.9

**Comment by PhysicallyActive** [comment, marked as answer]
You could solve your problem by using `Annotated` arguments to attach metadata to the body-parameters:

```python
from typing import Annotated, Optional

import fastapi
from pydantic import BaseModel, Field


class Param(BaseModel):
    param: str = Field(default=None)


PARAM = fastapi.Body(
    description="param",
    embed=True,
)


app = fastapi.FastAPI()


@app.post("/one")
async def post_one(one: Annotated[Optional[Param], PARAM] = None):
    return one


@app.post("/two")
async def post_two(two: Annotated[Optional[Param], PARAM] = None):
    return two
```

**NOTE**: you can't have a default value be set in the `Annotated` metadata (i.e., `= None` is required).

**Reply by tonyay163** [reply]
you're right, this seems to work, but this is definitely a regression. If we don't support the other method anymore, we should include that as a breaking change.
